### PR TITLE
Provide class for extension to insert sandboxed analytics

### LINKED ISSUE #9484: Provide APIs to 3p extensions for analytics custom events
Two API methods are provided for extension element to send analytics pings and customize their pings in different use cases.

### Manual trigger - CustomEventReporter class
Analytics triggers typically come from user events but one can also do manual triggering to send analytics pings with their customized data. 

To trigger custom events, 3P extension element first need to build a `customEventReporter` using `customEventReporterBuilder`.

For example
```
  buildCallback() {
   const builder = new customEventReporterBuilder(this.element);
   builder.track('layout', 'example.come/layout');
   builder.track('id', ['example.com/${id}', 'example2.com/${id}']);
   this.reporter_ = builder.build();
  }
```
`customEventReporterBuilder` constructor requires the parent AMP element. A builder instance provide two methods `track(eventName, request)` and `build()`. `#track()` takes in a custom event name, and the corresponding request (or an array of requests) to that event. While `#build()` should be called after adding all custom event, and will return the built `customEventReporter` instance.

The `customEventReporter` class provide a `trigger` method to manual trigger created custom events on 3P extension element with optional vars. 

For example
```
layoutCallback() {
  this.reporter_.trigger('layout');
  getIdPromise.then(id => {
    this.reporter_.trigger('id', {'id': '123'});
  }
}
```

Please note that with the use of `CustomEventReporter`, we assume all custom events are sandbox events that are scoped inside extension element. Thus `CustomEventReporter` can only used to report and collect 3P extension element created custom events. 

### Config 
Extensions can configure their [analytics config](https://github.com/ampproject/amphtml/blob/master/extensions/amp-analytics/amp-analytics.md#configuration) to get analytics events.

- To create the adapter class, extension element needs to provide a function that returns a promise with analytics config. 
```
const adapter = new SandboxAnalyticsAdapter(this.element, createConfigFunc);
```
The analytics config can be static, and checked in to the extension folder. For example:
```
import {analyticsConfig} from 'extensions/amp-***/0.1/analytics-config.js;
createConfigFunc = () => return Promise.resolve(analyticsConfig);
```
Or the analytics config can be fetched remotely, and available asynchronously.  For example:
```
createConfigFunc = fetchRemoteInfo.then(info => {
   return extractAnalyticsConfig(info);
};
```
- Restrictions apply to sandboxed analytics element's config. These Includes not supporting "type" or "config" attributes, limiting selector in parent element's scope, and applying variable replacement for whitelisted variables only.

### Behind the scene

`SandboxAnalyticsAdapter` will insert an `<amp-analytics>`  to DOM when its parent extension element starts layout. We delay the creation of analytics element because it's rare that an extension element collect analytics data before layout. And analytics is by default given lower priority than other contents. 

Those inserted amp-analytics elements are sandboxed in its parent extension element. This means it cannot access element outside its parent scope, and the [url variable substitution](https://github.com/ampproject/amphtml/blob/master/spec/amp-var-substitutions.md) only works if variable is included in the [whitelist](https://github.com/ampproject/amphtml/blob/master/extensions/amp-analytics/0.1/sandbox-vars-whitelist.js). 

The inserted `<amp-analytics>` element is not ready to handle analytics events until it has been initialized.  Before that `SandboxAnalyticsAdapter` class will buffer manual triggered events.  Please also note that the buffer has a size limit, and it only works for manual triggered events, not for other events happen before layout. 

If the extension doesn't [require relayout after unlayout](https://github.com/ampproject/amphtml/blob/master/src/base-element.js#L447), one can ignore this part. 
In the case of relayout, the adapter class will clean old buffered events, and wait for the new analytics element's ready signal. It is possible that elements can get laid out multiple times. One need to make sure the `createConfigFunc` returns the new  promise with new config after relayout.

### PR DESCRIPTION
**Pull request by zhouyx**
Closes #9484.

### CODE REVIEW

File: extensions/amp-analytics/0.1/amp-analytics.js [hunk 1]
```diff
@@ -186,7 +186,9 @@ export class AmpAnalytics extends AMP.BaseElement {
             .then(instrumentation => {
               this.instrumentation_ = instrumentation;
             })
-            .then(this.onFetchRemoteConfigSuccess_.bind(this));
+            .then(this.onFetchRemoteConfigSuccess_.bind(this))
+            // USE CommonSignals.LOAD_END once get rid of trigger = immediate
+            .then(() => this.signals().signal('amp-analytics-ready'));
```
Comment (lannka) [inline extensions/amp-analytics/0.1/amp-analytics.js]: put a TODO?

File: src/base-element.js [hunk 2]
```diff
@@ -369,6 +369,25 @@ export class BaseElement {
   }
 
   /**
+   * Subclasses can override this method to create an analytics config.
+   * @returns {?Promise<!JSONType>}
```
Comment (lannka) [inline src/base-element.js]: better to use `{!Promise<(!JSONType|undefined)>}`

File: src/base-element.js [hunk 3]
```diff
@@ -369,6 +369,25 @@ export class BaseElement {
...
+   */
+  createSandboxAnalyticsConfigCallback() {
+    return null;
+  }
+
+  /**
+   * Called on the inserted sandbox analytics element is ready
+   * Subclasses can override this method.
+   */
+  onSandboxAnalyticsReadyCallback() {}
```
Comment (lannka) [inline src/base-element.js]: how about 
`whenSandboxAnalyticsReady() { return readyPromise }`

File: src/base-element.js [hunk 4]
```diff
@@ -369,6 +369,25 @@ export class BaseElement {
...
+
+  /** Subclasses can override this method */
+  isloadingAnalyticsExtension() {
```
Comment (lannka) [inline src/base-element.js]: let's not expose this for now :-)

File: src/custom-element.js [hunk 5]
```diff
@@ -1716,6 +1730,27 @@ function createBaseCustomElementClass(win) {
       });
     }
 
+    insertSandboxAnalytics_() {
+      this.sandboxAnalyticsConfigPromise_ =
+          this.implementation_.createSandboxAnalyticsConfigCallback();
+      if (this.sandboxAnalyticsConfigPromise_) {
+        this.sandboxAnalyticsConfigPromise_.then(config => {
+          this.sandboxAnalyticsConfigPromise_ = null;
+          this.sandboxAnalyticsElement_ = insertAnalyticsElement(
```
Comment (dvoytenko) [inline src/custom-element.js]: How much extra are compiling in here?

Comment (zhouyx) [inline src/custom-element.js]: Sorry I didn't get the question...

Comment (dvoytenko) [inline src/custom-element.js]: This is a part of core. Using `insertAnalyticsElement` would to the core binary.

Comment (zhouyx) [inline src/custom-element.js]: I see your point. But `insertAnalyticsElement` was in core binary before, in `src/analytics.js`.
FYI I get the size:
master: 
  1.14 MB   |   217.61 kB   |                             65.44 kB   |   v0.js / amp.js
After this PR:
  1.13 MB   |   218.48 kB   |                             65.64 kB   |   v0.js / amp.js

File: src/base-element.js (hunk 2 above)
Comment (zhouyx) [inline src/base-element.js]: Why is the other way better?  I like `{?Promise<!JSONType>}` because it saves one promise call and returns a bit earlier

File: src/base-element.js (hunk 3 above)
Comment (zhouyx) [inline src/base-element.js]: return a promise is probably better, but it's even more confusing for element (like ad) that can relayout. Note: `onSandboxAnalyticsReadyCallback()` can be called multiple times based on the number of layout times of an element.

**Comment by zhouyx** [comment]
Friendly ping : )

**Comment by zhouyx** [comment]
👀 PTAL

**Comment by zhouyx** [comment]
As discussed offline, move the logic outside customElement class, have a class that baseElement can import instead. 
PTAL

File: src/custom-element.js [hunk 6]
```diff
@@ -599,6 +599,11 @@ function createBaseCustomElementClass(win) {
         delete this[dom.UPGRADE_TO_CUSTOMELEMENT_RESOLVER];
         delete this[dom.UPGRADE_TO_CUSTOMELEMENT_PROMISE];
       }
+      /** @private {?Promise<!JSONType>} */
+      this.sandboxAnalyticsConfigPromise_ = null;
+
+      /** @private {?Element} */
+      this.sandboxAnalyticsElement_ = null;
```
Comment (jridgewell) [inline src/custom-element.js]: Why are these added?

File: src/analytics.js [hunk 7]
```diff
@@ -105,3 +110,107 @@ export function insertAnalyticsElement(
...
+  /**
+   * Private function that get config and insert analytics element
+   */
+  createSandboxAnalyticsElement_() {
+    if (this.analyticsElement || this.getConfigPromise_) {
+      return;
+    }
+    this.getConfigPromise_ = this.createConfigFunc_();
+    this.getConfigPromise_.then(config => {
+      if (!this.getConfigPromise_) {
```
Comment (jridgewell) [inline src/analytics.js]: If we can unload then reload, there's a chance that a new promise is here. Check that they are the same instance.

Comment (zhouyx) [inline src/analytics.js]: good point

**Comment by zhouyx** [comment]
I updated two sample use cases. Will continue on it once we feel comfortable with the design.

**Comment by zhouyx** [comment]
@lannka PTAL

File: src/analytics.js [hunk 8]
```diff
@@ -106,3 +111,39 @@ export function insertAnalyticsElement(
...
   return analyticsElem;
 }
+
+/**
+ * A function that is used by extension element to get analytics events.
+ * The function makes sure to create or remove sandbox amp-analytics element at right time.
+ * @param {!AmpElement} parent
+ * @param {!Promise<!JsonObject>} promise
+ */
+export function getAnalyticsInSandbox(parent, promise) {
```
Comment (lannka) [inline src/analytics.js]: can you rename? it does not return anything, why it's called "get"

File: src/analytics.js [hunk 9]
```diff
@@ -106,3 +111,39 @@ export function insertAnalyticsElement(
   parentElement.appendChild(analyticsElem);
   return analyticsElem;
 }
+
+/**
+ * A function that is used by extension element to get analytics events.
+ * The function makes sure to create or remove sandbox amp-analytics element at right time.
+ * @param {!AmpElement} parent
+ * @param {!Promise<!JsonObject>} promise
```
Comment (lannka) [inline src/analytics.js]: `promise` -> `configPromise`

File: src/analytics.js [hunk 10]
```diff
@@ -106,3 +111,39 @@ export function insertAnalyticsElement(
   parentElement.appendChild(analyticsElem);
   return analyticsElem;
 }
+
+/**
+ * A function that is used by extension element to get analytics events.
+ * The function makes sure to create or remove sandbox amp-analytics element at right time.
+ * @param {!AmpElement} parent
```
Comment (lannka) [inline src/analytics.js]: `parent` -> `element`

File: src/analytics.js [hunk 11]
```diff
@@ -106,3 +111,39 @@ export function insertAnalyticsElement(
   parentElement.appendChild(analyticsElem);
   return analyticsElem;
 }
+
+/**
+ * A function that is used by extension element to get analytics events.
```
Comment (lannka) [inline src/analytics.js]: Suggestion: 
```
A helper method that should be used by all extension elements to add their sandbox analytics tracking. This method takes care of insert and remove the analytics tracker at the right time of the element lifecycle.
```

Comment (zhouyx) [inline src/analytics.js]: changed

File: src/analytics.js (hunk 10 above)
Comment (zhouyx) [inline src/analytics.js]: done

File: src/analytics.js (hunk 8 above)
Comment (zhouyx) [inline src/analytics.js]: renamed to `useAnalyticsInSandbox`

File: src/analytics.js (hunk 9 above)
Comment (zhouyx) [inline src/analytics.js]: I kept this unchanged, because I want the promise here to be type `{!Promise<!JsonObject>}` but I will need to set it to `promise = null` after unlayout to prevent false adding element if promise resolve after unlayout. It would require me to create an extra param to store the state anyway if I force the input promise to by not null.
